### PR TITLE
fix(inventory): prevent SQL error when handle manufacturer rule

### PR DESCRIPTION
**Pull request by stonebuzz**
Prevent SQL error 
```sql
Incorrect integer value: 'Microsoft' for column glpi_softwares.manufacturers_id at row 1 in src\DBmysql.php at line 1969
```

1. To fix SQL : fix Bad use of ```$this->known_links``` from ```Software```, ```key``` are never  use ```md5```
2. this fix caused a problem with the management of the operating system of the software version 

```OperatingSystem``` is a ```CommonDropdown``` and ```InventoryAsset->handleLinks``` with ```Dropdown::importExternal``` return``` -1``` if name is ```0```

Because 

```CommonDropdown::importExternal``` check if ```strlen($value) == 0``` (true if this case)
then call ```CommonDropdown::importExternal``` which check if ```empty($value)``` (false if this case and return ```-1```)

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #13925 [#301](https://github.com/glpi-project/glpi-inventory-plugin/issues/301)

### CODE REVIEW

File: src/CommonDropdown.php [hunk 1]
```diff
@@ -743,7 +743,7 @@ public function importExternal(
     ) {
 
         $value = trim($value);
-        if (strlen($value) == 0) {
+        if (empty($value)) {
```
Comment (cedric-anne) [inline src/CommonDropdown.php:746]: This change seems dangerous to me.
Is it possible to filter `0` values on inventory side ?

Comment (trasher) [reply src/CommonDropdown.php:746]: Why should that be dangerous?

Comment (trasher) [reply src/CommonDropdown.php]: CommonDropdown::import() that is finally called use empty on provided name, I really do not undesratnd why importExternal should only check empty string

**Comment by trasher** [comment]
As far as I've seen, two ocurrences has been missed:

```diff
diff --git a/src/Inventory/Asset/Software.php b/src/Inventory/Asset/Software.php
index 27546a3b46..9796b078fd 100644
--- a/src/Inventory/Asset/Software.php
+++ b/src/Inventory/Asset/Software.php
@@ -375,7 +375,7 @@ class Software extends InventoryAsset
 
             //update softwarecategories if needed
             //reconciles the software without the version (no needed here)
-            $sckey = 'softwarecategories_id' . ($val->softwarecategories_id ?? 0);
+            $sckey = md5('softwarecategories_id' . ($val->softwarecategories_id ?? 0));
             if (
                 isset($db_software_data[$key_wo_version])
                 && $db_software_data[$key_wo_version]['softwarecategories'] != ($this->known_links[$sckey] ?? 0)
@@ -578,7 +578,7 @@ class Software extends InventoryAsset
                 continue;
             }
 
-            $mkey = 'manufacturers_id' . $val->manufacturers_id;
+            $mkey = md5('manufacturers_id' . $val->manufacturers_id);
             $input = Sanitizer::encodeHtmlSpecialCharsRecursive([
                 'name'             => $val->name,
                 'manufacturers_id' => $this->known_links[$mkey] ?? 0
```

**Comment by stonebuzz** [comment]
To clarify 

Currently the software inventory manages manages on its side ```manufacturers_id``` (and ```softwarecategories_id```)  
because these data are necessary during the reconciliation (to know if it is necessary to update / insert / delete)

This management is done with the ```$this->known_links``` and  ```key``` **without** ```md5()```

```php
    //If the manufacturer has been modified or set by the rules engine
    if (isset($res_rule["manufacturer"])) {
        $mkey = 'manufacturers_id' . $res_rule['manufacturer'];
        $mid = Dropdown::import(
            'Manufacturer',
            ['name' => $res_rule['manufacturer']]
        );
        $this->known_links[$mkey] = $mid;
        $val->manufacturers_id = $res_rule['manufacturer'];
    } else if (
        property_exists($val, 'manufacturers_id')
        && $val->manufacturers_id != ''
        && $val->manufacturers_id != '0'
    ) {
        $mkey = 'manufacturers_id' . $val->manufacturers_id;
        if (!isset($this->known_links[$mkey])) {
            $new_value = Dropdown::importExternal(
                'Manufacturer',
                addslashes($val->manufacturers_id),
                $this->entities_id
            );
            $this->known_links[$mkey] = $new_value;
        }
    } else {
        $val->manufacturers_id = 0;
    }
```

The problem occurs when the user uses the manufacturers' dictionary

From ```src/Inventory/Asset/Software.php``` ```RuleDictionnarySoftwareCollection``` is used.
But from  ```InventoryAsset->handleLinks()``` a dedicated code for ```manufacturers_id``` allow user to use ```RuleDictionnaryManufacturerCollection``` 

```php
    if ($key == "manufacturers_id" || $key == 'bios_manufacturers_id') {
        $manufacturer = new Manufacturer();
        unset($this->raw_links[$known_key]);
        $val  = $manufacturer->processName($val);
        $known_key = md5($key . $val);
        //keep raw values...
        $this->raw_links[$known_key] = $val;
        if ($key == 'bios_manufacturers_id') {
            $foreignkey_itemtype[$key] = getItemtypeForForeignKeyField('manufacturers_id');
        }
    }
```

here the new value is kept in memory in  ```$this->known_links``` and  ```key```  **with** ```md5()```

When a ```Software``` is manage, ```Software->cleanInputToPrepare()``` can't get new value from ```InventoryAsset->handleLinks()``` because of ```md5()``` and return ```raw``` value.

At this step GLPI trigger an error 

```sql
Incorrect integer value: 'Microsoft' for column glpi_softwares.manufacturers_id at row 1 in src\DBmysql.php at line 1969
```

To prevent this SQL error, we need to use ```md5()``` key .

But this change revealed another problem.

```InventoryAsset->handleLinks()``` use ```Dropdown::importExternal``` for ```CommonDropdown::class``` without checking if ```$value->$key```.

At the end if ```$value->$key = 0 ```, ```Dropdown::importExternal``` return ```-1```and another SQL error is triggered

```php
PHP User Error (256): Out of range value for column 'operatingsystems_id' at row 1 in /home/stanislas/TECLIB/DEV/GLPI/10.0-bugfixes/src/DBmysql.php at line 1969
```

To prevent this, I added a check on the value , to be sure not to insert ```0```

**Review by cedric-anne — APPROVED** [review]

**Review by cedric-anne — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Comment by Steve-Hubert** [comment]
Hello @stonebuzz 
I encountered this bug after updating GLPI to 10.0.6.
After edit the files with commits in this PR, I encounter a new error that seems related:

_php-errors.log_
```
[2023-02-02 15:35:42] glpiphplog.ERROR:   *** PHP User Error (256): Out of range value for column 'manufacturers_id' at row 1 in /var/www/html/glpi/src/DBmysql.php at line 1969
  Backtrace :
  src/DBmysql.php:1969                               trigger_error()
  src/Inventory/Asset/Software.php:714               DBmysql->executeStatement()
  src/Inventory/Asset/Software.php:443               Glpi\Inventory\Asset\Software->storeSoftware()
  src/Inventory/Asset/MainAsset.php:871              Glpi\Inventory\Asset\Software->handle()
  src/Inventory/Asset/MainAsset.php:791              Glpi\Inventory\Asset\MainAsset->handleAssets()
  src/RuleImportAsset.php:988                        Glpi\Inventory\Asset\MainAsset->rulepassed()
  src/Rule.php:1522                                  RuleImportAsset->executeActions()
  src/RuleCollection.php:1626                        Rule->process()
  src/Inventory/Asset/MainAsset.php:571              RuleCollection->processAllRules()
  src/Inventory/Inventory.php:706                    Glpi\Inventory\Asset\MainAsset->handle()
  src/Inventory/Inventory.php:341                    Glpi\Inventory\Inventory->handleItem()
  src/Inventory/Request.php:360                      Glpi\Inventory\Inventory->doInventory()
  src/Inventory/Request.php:90                       Glpi\Inventory\Request->inventory()
  src/Agent/Communication/AbstractRequest.php:359    Glpi\Inventory\Request->handleAction()
  src/Agent/Communication/AbstractRequest.php:271    Glpi\Agent\Communication\AbstractRequest->handleJSONRequest()
  front/inventory.php:92                             Glpi\Agent\Communication\AbstractRequest->handleRequest()
```

**Comment by stonebuzz** [comment]
@Steve-Hubert  can you retry today with nightly-build to be sure that you have modifications

This problem is normally solved with  https://github.com/glpi-project/glpi/pull/13930/commits/cbbb768461e83fc241de5a8712c5aa7ae61306e5

Open issue if problem persist.

**Pull request merged**